src/flash.rs

```diff
@@ -219,12 +208,14 @@ impl<'a> FlashWriter<'a> {
 
         let address = (FLASH_START + offset) as *const _;
 
-        Ok(FlashReaderGuard {
-            _flash: self,
-
-            // NOTE(unsafe) read with no side effects safe-guarded by FlashReaderGuard.
-            data: unsafe { core::slice::from_raw_parts::<'static, u8>(address, length) },
-        })
+        Ok(
+            // NOTE(unsafe) read with no side effects. The data returned will
+            // remain valid for its lifetime because we take an immutable
+            // reference to this FlashWriter, and any operation that would
+            // invalidate the data returned would first require taking a mutable
+            // reference to this FlashWriter.
+            unsafe { core::slice::from_raw_parts::<'static, u8>(address, length) },
```

Comment:
This shouldn't be `'static`, it should have the same lifetime of `&self`. Not sure if you need to strictly annotate it in order for the compiler to be happy.

Reply:
It looks like the compiler can infer the lifetime, as this compiles:
```rust
 unsafe { core::slice::from_raw_parts(address, length) }
```
However, I decided to explicitly annotate the lifetimes around `read()` to make the relationship more obvious.

Reply:
Hmm, but I don't think the `'a` lifetime was the best idea, since it's the same lifetime as `Parts` like in the other case.
To be honest I think we should just let the compiler do the job and use just 
```rust
pub fn read(&self, offset: u32, length: usize) -> Result<&[u8]> {
    //...
    unsafe { core::slice::from_raw_parts(address, length) }
}
```

The lifetime elisions rules seem to be very clear in this case anyway, so it's best to leave the compiler to choose the appropriate lifetime I think. Sorry for all the nitpicking .-.

Reply:
Haha don't be sorry, glad to get feedback! I didn't realize annotating the lifetime would change the behavior, I thought I was just making something that was implicit explicit.

Reply:
It was just because it was the same lifetime you used at
```rust
impl<'a> FlashWriter<'a>
```